src/states/proving_node.rs

```diff
@@ -0,0 +1,649 @@
+                    self.handle_connection_info_request(
+                        encrypted_conn_info,
+                        pub_id,
+                        msg_id,
+                        msg.src,
+                        msg.dst,
+                        outbox,
+                    )?
+                } else {
+                    self.add_message_to_backlog(RoutingMessage {
+                        content: ConnectionInfoRequest {
+                            encrypted_conn_info,
+                            pub_id,
+                            msg_id,
+                        },
+                        ..msg
+                    })
+                }
+            }
+            RoutingMessage {
+                content:
+                    ConnectionInfoResponse {
+                        encrypted_conn_info,
+                        pub_id,
+                        msg_id,
+                    },
+                src: ManagedNode(src_name),
+                dst: Client { .. },
+            } => self.handle_connection_info_response(
+                encrypted_conn_info,
+                pub_id,
+                msg_id,
+                src_name,
+                msg.dst,
+            )?,
+            RoutingMessage {
+                content: NodeApproval(gen_info),
+                src: PrefixSection(_),
+                dst: Client { .. },
+            } => return Ok(self.handle_node_approval(gen_info)),
+            RoutingMessage {
+                content: Ack(ack, _),
+                ..
+            } => self.handle_ack_response(ack),
+            _ => {
+                self.add_message_to_backlog(msg);
+            }
+        }
+
+        Ok(Transition::Stay)
+    }
+
+    // Backlog the message to be processed once we are approved.
+    fn add_message_to_backlog(&mut self, msg: RoutingMessage) {
+        trace!(
+            "{} Not approved yet. Delaying message handling: {:?}",
+            self,
+            msg
+        );
+        self.msg_backlog.push(msg);
+    }
+
+    fn handle_node_approval(&mut self, gen_pfx_info: GenesisPfxInfo) -> Transition {
+        self.resource_prover.handle_approval();
+        info!(
+            "{} Resource proof challenges completed. This node has been approved to join the \
+             network!",
+            self
+        );
+
+        Transition::IntoNode { gen_pfx_info }
+    }
+
+    fn handle_ack_response(&mut self, ack: Ack) {
+        self.ack_mgr.receive(ack)
+    }
+
+    fn handle_timeout(&mut self, token: u64, outbox: &mut EventBox) -> Transition {
+        let log_ident = format!("{}", self);
+        if let Some(transition) = self
+            .resource_prover
+            .handle_timeout(token, log_ident, outbox)
+        {
+            transition
+        } else {
+            self.resend_unacknowledged_timed_out_msgs(token);
+            Transition::Stay
+        }
+    }
+
+    fn send_connection_info_request(
```

Comment:
This is mostly a copy paste of the one still in src/states/node.rs.
We probably should share the bulk of the implementation.

Is there other like these, I'll try to point them out.

Reply:
You are right. I'll try to extract the common parts and put them into `Relocated`.